L1Trigger/Phase2L1Taus/src/L1HPSPFTauBuilder.cc

```diff
@@ -0,0 +1,553 @@
+
+L1HPSPFTauBuilder::L1HPSPFTauBuilder(const edm::ParameterSet& cfg)
+  : signalConeSizeFormula_(nullptr)
+  , min_signalConeSize_(cfg.getParameter<double>("min_signalConeSize"))
+  , max_signalConeSize_(cfg.getParameter<double>("max_signalConeSize"))
+  , useStrips_(cfg.getParameter<bool>("useStrips"))
+  , stripSize_eta_(cfg.getParameter<double>("stripSize_eta"))
+  , stripSize_phi_(cfg.getParameter<double>("stripSize_phi"))
+  , isolationConeSize_(cfg.getParameter<double>("isolationConeSize"))
+  , inputFileName_rhoCorr_(cfg.getParameter<std::string>("inputFileName_rhoCorr"))
+  , inputFile_rhoCorr_(nullptr)
+  , histogramName_rhoCorr_(cfg.getParameter<std::string>("histogramName_rhoCorr"))
+  , histogram_rhoCorr_(nullptr)  
+  , histogram_rhoCorr_yMax_(-1.)
+  , debug_(cfg.getUntrackedParameter<bool>("debug", false))
+{
+  std::string signalConeSizeFormulaName = Form("signalConeSizeFormula%i", signalConeSizeFormula_instance_counter_);
+  ++signalConeSizeFormula_instance_counter_;
+  std::string signalConeSizeFormula_original = cfg.getParameter<std::string>("signalConeSize");
+  TString signalConeSizeFormula_modified = signalConeSizeFormula_original.data();
+  signalConeSizeFormula_modified.ReplaceAll("pt", "x");
+  signalConeSizeFormula_ = new TFormula(signalConeSizeFormulaName.data(), signalConeSizeFormula_modified.Data(), false, false);
+  if ( !(signalConeSizeFormula_->Compile() == 0) )
+  {
+    throw cms::Exception("L1HPSPFTauBuilder")
+      << "Invalid Configuration parameter 'signalConeSize' = '" << signalConeSizeFormula_original << "' !!\n";
+  }
+  assert(max_signalConeSize_ >= min_signalConeSize_);
+
+  isolationConeSize2_ = isolationConeSize_*isolationConeSize_;
+
+  if ( debug_ )
+  {
+    std::cout << "setting Quality cuts for signal PFCands:" << std::endl;
+  }
+  edm::ParameterSet cfg_signalQualityCuts = cfg.getParameter<edm::ParameterSet>("signalQualityCuts");
+  signalQualityCuts_dzCut_disabled_           = readL1PFTauQualityCuts(cfg_signalQualityCuts,    "disabled",        debug_);
+  signalQualityCuts_dzCut_enabled_primary_    = readL1PFTauQualityCuts(cfg_signalQualityCuts,    "enabled_primary", debug_);
+  if ( debug_ )
+  {
+    std::cout << "setting Quality cuts for isolation PFCands:" << std::endl;
+  }
+  edm::ParameterSet cfg_isolationQualityCuts = cfg.getParameter<edm::ParameterSet>("isolationQualityCuts");
+  isolationQualityCuts_dzCut_disabled_        = readL1PFTauQualityCuts(cfg_isolationQualityCuts, "disabled",        debug_);
+  isolationQualityCuts_dzCut_enabled_primary_ = readL1PFTauQualityCuts(cfg_isolationQualityCuts, "enabled_primary", debug_);
+  isolationQualityCuts_dzCut_enabled_pileup_  = readL1PFTauQualityCuts(cfg_isolationQualityCuts, "enabled_pileup",  debug_);
+
+  if ( inputFileName_rhoCorr_ != "" && histogramName_rhoCorr_ != "" ) 
+  {
+    LocalFileInPath inputFileName_rhoCorr_fip(inputFileName_rhoCorr_);
+    inputFile_rhoCorr_ = openFile(inputFileName_rhoCorr_fip);
```

Comment:
Can you wrap this in a `std::unique_ptr<TFile>`, so you don't need the explicit `delete` below ?

Alternatively, the `openFile()` function could be modified to return a `std::unique_ptr<TFile>`.